src/Smalot/PdfParser/PDFObject.php

```diff
@@ -224,6 +224,65 @@ private function formatContent(?string $content): string
             return '';
         }
 
+        // Find all inline image content and replace them so they aren't
+        // affected by the next steps
+        $pdfInlineImages = [];
+        $offsetBI = 0;
+        while (preg_match('/\sBI\s(\/.+?)\sID\s(.+?)\sEI(?=\s|$)/s', $content, $text, \PREG_OFFSET_CAPTURE, $offsetBI)) {
+            // Attempt to detemine if this instance of the 'BI' command
+            // actually occured within a (string) using the following
+            // steps:
+
+            // Remove any escaped parentheses from the alleged image
+            // characteristics data
+            $para = str_replace(['\\(', '\\)'], '', $text[1][0]);
+
+            // Remove all correctly ordered and balanced parentheses
+            // from (strings)
+            do {
+                $paraTest = $para;
+                $para = preg_replace('/\(([^)]*)\)/', '$1', $paraTest);
+            } while ($para != $paraTest);
+
+            $paraOpen = strpos($para, '(');
+            $paraClose = strpos($para, ')');
+
+            // If the remaining text contains a close parenthesis ')'
+            // AND it occurs before any open parenthesis, then we are
+            // almost certain to be inside a (string)
+            if (0 < $paraClose && (false === $paraOpen || $paraClose < $paraOpen)) {
+                // Bump the search offset forward and match again
+                $offsetBI = (int) $text[1][1];
+                continue;
+            }
+
+            // Double check that this is actually inline image data by
+            // parsing the alleged image characteristics as a dictionary
+            $dict = $this->parseDictionary('<<'.$text[1][0].'>>');
+
+            // Check if an image Width and Height are set in the dict
+            if ((isset($dict['W']) || isset($dict['Width']))
+                && (isset($dict['H']) || isset($dict['Height']))) {
+                $id = uniqid('IMAGE_', true);
+                $pdfInlineImages[$id] = [
+                    preg_replace(['/\r\n/', '/\r/', '/\n/'], ' ', $text[1][0]),
+                    preg_replace(['/\r\n/', '/\r/', '/\n/'], '', $text[2][0]),
+                ];
+                $content = preg_replace(
+                    '/'.preg_quote($text[0][0], '/').'/',
+                    '^^^'.$id.'^^^',
+                    $content,
+                    1
+                );
+            } else {
+                // If there was no valid dictionary, or a height and width
+                // weren't specified, then we don't know what this is, so
+                // just leave it alone; bump the search offset forward and
+                // match again
+                $offsetBI = (int) $text[1][1];
```

Comment:
Maybe something like the following would prevent array-access errors (to somet extent)?

```suggestion
                $offsetBI = (int) $text[1][1] ?? 1;
```

The `?? 1` is some dummy value, it should be replaced with a default value which makes sense, but only if there is a chance that `$text[1][1]` can be undefined.

Reply:
I think this is a decent idea in theory, but it couldn't be implemented as simply as this.

First of all I don't think there is a case where asking for `preg_match` to return offset values returns successful matches, but *without* the offset values. So is it even necessary?

Second, if you wanted to add it, you couldn't just use a static value because of the potential for falling into an endless loop. The next iteration of the loop would find the same match again, have the offset set to `1` and find the same match again, etc. So if the offset value was somehow missing, you would have to, say, take the first 10 or 20 characters of the match and match it against the document stream to find the offset, then use *that* as the fallback.

I believe it's probably too much trouble to account for, for something that perhaps might never happen.